### ISSUE TITLE
Optimize speed

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/cpp-test-runner#12]
The C++ test runner is functional right now, but it is quite slow. In fact, it is so slow that the test runner times out before being able to return the results. We should speedup the test runner to not have it timeout.

**Comment by iHiD** [comment]
(cc @exercism/cpp)

**Comment by neenjaw** [comment]
Wonder if you can pre-build the cpp test package (and maybe also each exercise's test suite) so that during the test running time, all you have to do is compile the solution, then just link it with the pre-built assets.

**Comment by silvanocerza** [comment]
> Wonder if you can pre-build the cpp test package (and maybe also each exercise's test suite) so that during the test running time, all you have to do is compile the solution, then just link it with the pre-built assets.

Yes, this can technically be done. 

We'd have to create a `catch_main.cpp` to declare `main`, build it as a library and add it as library dependency when building the test.

**Comment by silvanocerza** [comment]
I took a look around. As of now we're building Catch as a library only when running automated tests on the C++ track, mainly to speed them up otherwise we'd have the same issue we have here, each test would recompile the whole library.

I think the best way to go with this would be to build the Catch library directly inside the Docker image so it's available in the test runner, we'll also need to update all exercises' `CMakeLists.txt` to find Catch as a library; as of now it can't be done without edits since detection of Catch in an exercise's `CMakeLists.txt` requires that the `catchlib` target is defined in a super project `CMakeLists.txt`. 

See: 
https://github.com/exercism/cpp/blob/837223df2591289616a81793d22ac5b6e1eca685/exercises/concept/strings/CMakeLists.txt#L30

Is this feasible?

**Comment by ErikSchierboom** [comment]
@silvanocerza As all the existing `CMakeLists.txt` files already contain the `if(EXERCISM_COMMON_CATCH)`, do we even need to update the existing exercises? It sounds like we could update the Dockerfile to build the Catch library in advance, and then we could modify the `bin/run.sh` file to pass the `-DEXERCISM_COMMON_CATCH=1` option when calling cmake (see https://github.com/exercism/cpp-test-runner/blob/main/bin/run.sh#L43), right?

If you have Docker installed, you can verify that everything still works by running `./bin/run-tests-in-docker.sh`. If there are no error messages, all is good (CI will verify this too).

As I'm off for the next 10 days, it would be great if someone could PR this change to get C++ in the beta with a working test runner! :)

**Comment by silvanocerza** [comment]
> @silvanocerza As all the existing CMakeLists.txt files already contain the if(EXERCISM_COMMON_CATCH), do we even need to update the existing exercises? It sounds like we could update the Dockerfile to build the Catch library in advance, and then we could modify the bin/run.sh file to pass the -DEXERCISM_COMMON_CATCH=1 option when calling cmake (see https://github.com/exercism/cpp-test-runner/blob/main/bin/run.sh#L43), right?

That cannot work, we use the CMake generator expression [`$<TARGET_OBJECTS:catchlib>`](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) to get the list of objects built from the target `catchlib`, problem is we never define it here.

That option is only used in the main `CMakeLists.txt` of the `cpp` track: https://github.com/exercism/cpp/blob/837223df2591289616a81793d22ac5b6e1eca685/CMakeLists.txt#L24-L31

Just adding `-DEXERCISM_COMMON_CATCH=1` to `bin/run.sh` results in the build failing as expected:

```-- The CXX compiler identification is GNU 9.3.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
CMake Error at CMakeLists.txt:23 (add_executable):
  Error evaluating generator expression:

    $<TARGET_OBJECTS:catchlib>

  Objects of target "catchlib" referenced but no such target exists.


CMake Error at CMakeLists.txt:23 (add_executable):
  No SOURCES given to target: example-syntax-error


CMake Generate step failed.  Build files cannot be regenerated correctly. 
/opt/test-runner
```

**Comment by patricksjackson** [comment]
Building the Catch library in advance and including it in the Docker image is the way to go. (Assuming that that work is outside the critical path, I'm not entirely certain what all is included in the path that is too slow here. Catch should only be taking ~6 seconds to build so if that's what's causing the timeout then the timing already seems to be pretty tight. I know that naive solutions to some of the practice exercises should also timeout if 6 seconds extra is too much.)

Without modifying the exercise CMake files, you'll need to have a root CMakeLists.txt file to define the library and call out to each exercise as a subproject. And then in the Dockerfile build a dummy hello world project to force the building of the object files.

The existing root CMakeLists.txt file is probably very close to what you need. It probably just needs to be modified to use the directory structure here.

**Comment by iHiD** [comment]
So for context:
- We have a 10s timeout window. 
- Most languages are in the 0.5-1.5s range. 
- The slowest is about 6s to compile (in line with what you're saying here). 
 
So I guess c++ is just very slow to compile? Which is just a bit unexpected for me.

**Comment by patricksjackson** [comment]
C++ is very slow to compile, it's a very common issue and in fact some of the recent changes to the language are intended to help address it a little bit.

The decisions we made to use Catch for the testing library makes this worse too. We chose to switch to Catch because it was difficult to support other, precompiled libraries across many different platforms. C++ doesn't have a standard packaging system, so we had to rely on users figuring out how to install the right versions for their platform.

By contrast, Catch is simply the source code for the testing library in a heavily templated header file (which makes the compilation times pretty bad).  Catch actually has [a whole README page](https://github.com/catchorg/Catch2/blob/devel/docs/slow-compiles.md) dedicated to explaining how to speed up their slow compilation.

The actual tests and student code should compile and link with the Catch testing lib relatively quickly if we precompile Catch. I think **at best it will probably take 2-3 seconds**, so it will still be slower than most languages.

**Comment by iHiD** [comment]
2-3 seconds would be fine - well within norms for other compiled languages 🙂 

Is this something you could PR?

**Comment by ErikSchierboom** [comment]
@patricksjackson Would you be willing to submit a PR? I can help with the Docker bit if needed.

**Comment by iHiD** [comment]
Would anyone in this thread like to take this on, before we reach out to other channels to try and get someone to help? 🙂

**Comment by ErikSchierboom** [comment]
Closed by #17
